firmware/thruna/DCM/Src/App/states/App_DriveState.c

```diff
@@ -64,14 +67,29 @@ static void DriveStateRunOnTick1Hz(struct StateMachine *const state_machine)
 static void DriveStateRunOnTick100Hz(struct StateMachine *const state_machine)
 {
     // All states module checks for faults, and returns whether or not a fault was detected.
-    const bool all_states_ok    = App_AllStatesRunOnTick100Hz(state_machine);
-    const bool start_switch_off = !App_IsStartSwitchOn();
-    const bool bms_not_in_drive = !App_IsBmsInDriveState();
-    bool       exit_drive       = !all_states_ok || start_switch_off || bms_not_in_drive;
+    const bool all_states_ok        = App_AllStatesRunOnTick100Hz(state_machine);
+    const bool start_switch_off     = !App_IsStartSwitchOn();
+    const bool bms_not_in_drive     = !App_IsBmsInDriveState();
+    bool       exit_drive           = !all_states_ok || start_switch_off || bms_not_in_drive;
+    bool       regen_switch_enabled = App_IsRegenOn();
+    apps_pedal_percentage           = App_CanRx_FSM_PappsMappedPedalPercentage_Get();
+
+    // regen switched pedal percentage from [0, 100] to [-100, 100]
+    if (regen_switch_enabled)
+    {
+        apps_pedal_percentage = (apps_pedal_percentage - 50.0f) * 2;
+    }
 
     if (all_states_ok)
     {
-        App_SetPeriodicCanSignals_TorqueRequests();
+        if (apps_pedal_percentage < 0)
+        {
+            App_Run_Regen(&prev_torque_request, apps_pedal_percentage);
+        }
+        else
+        {
+            App_SetPeriodicCanSignals_TorqueRequests();
+        }
     }
 
     if (exit_drive)
```

Comment:
can we make `all_states_ok` and `exit_drive` the same condition? I don't see a condition where the start switch is off, or the bms is not in drive and we want to send one final torque request. I would prefer to see
```cpp
if(exit_drive) {
    App_SharedStateMachine_SetNextState(state_machine, App_GetInitState());
    return;
}

//map pedal and send torque request
```